README.md
# pelutils

Various utilities useful for Python projects. Features include

- Feature-rich logger using `Rich` for colourful printing
- Parsing for combining config files and command-line arguments - especially useful for parametric methods
- Time taking and profiling
- Easy to use data storage class for easy data saving and loading
- Table formatting
- Miscellaneous standalone functions providing various functionalities - see `pelutils/__init__.py`
- Data-science submodule with extra utilities for statistics, plotting, and machine learning using `PyTorch`
- `unique` function similar to `np.unique` but in linear time (currently Linux x86_64 only)

`pelutils` supports Python 3.7+.

[![pytest](https://github.com/peleiden/pelutils/actions/workflows/pytest.yml/badge.svg?branch=master)](https://github.com/peleiden/pelutils/actions/workflows/pytest.yml)
[![Coverage Status](https://coveralls.io/repos/github/peleiden/pelutils/badge.svg?branch=master)](https://coveralls.io/github/peleiden/pelutils?branch=master)

## Logging

Easy to use logger which fits common needs.

```py
log("This is printed but not saved to a log file as the logger has not been configured")

# Configure logger for the script
log.configure("path/to/save/log.log", "Optional title of log")

# Start logging
for i in range(70):  # Nice
    log("Execution %i" % i)

# Sections
log.section("New section in the logfile")

# Adjust logging levels
log.warning("Will be logged")
with log.level(LogLevels.ERROR):  # Only log at ERROR level or above
    log.warning("Will not be logged")
with log.no_log():
    log.section("I will not be logged")

# Error handling
# The zero-division error and stacktrace is logged
with log.log_errors:
    0 / 0
# Entire chained stacktrace is logged
with log.log_errors:
    try:
        0 / 0
    except ZeroDivisionError as e:
        raise ValueError("Denominator must be non-zero") from e

# User input - acts like built-in input but logs both prompt and user input
inp = log.input("WHAT... is your favourite colour? ")

# Log all logs from a function at the same time
# This is especially useful when using multiple threads so logging does not get mixed up
def fun():
    log("Hello there")
    log("General Kenobi!")
with mp.Pool() as p:
    p.map(collect_logs(fun), args)
```

## Time Taking and Profiling

Simple time taker inspired by Matlab Tic, Toc, which also has profiling tooling.

```py
TT.tick()
<some task>
seconds_used = TT.tock()

for i in range(100):
    TT.profile("Repeated code")
    <some task>
    TT.profile("Subtask")
    <some subtask>
    TT.end_profile()
    TT.end_profile()
print(TT)  # Prints a table view of profiled code sections

# Alternative syntax using with statement
with TT.profile("The best task"):
    <some task>

# Profile a loop
# Do not do this if the loop may be ended by a break statement!
for elem in TT.profile_iter(range(100), "The second best task"):
    <some task>

# When using multiprocessing, it can be useful to simulate multiple hits of the same profile
with mp.Pool() as p, tt.profile("Processing 100 items on multiple threads", hits=100):
    p.map(100 items)
```

## Data Storage

A data class that saves/loads its fields from disk.
Anything that can be saved to a `json` file will be.
Other data types will be saved to relevant file formats.

```py
@dataclass
class Person(DataStorage):
    name: str
    age: int
    numbers: np.ndarray

    json_name = "yoda.json"

yoda = Person(name="Yoda", age=900, numbers=np.array([69, 420]))
yoda.save("high/on/ketamine")  # Save to 'old' folder
# Saved data at high/on/ketamine/yoda.json
# {
#     "name": "Yoda",
#     "age": 900
# }
# There will also be a file named numbers.npy
yoda = Person.load("old")
```

## Parsing

A combination of parsing CLI and config file arguments which allows for a powerful, easy-to-use workflow.
Useful for parametric methods such as machine learning.

A file `main.py` could contain:
```py
options = {
    "learning-rate": { "default": 1.5e-3, "help": "Controls size of parameter update", "type": float },
    "gamma": { "default": 1, "help": "Use of generator network in updating", "type": float },
    "initialize-zeros": { "help": "Whether to initialize all parameters to 0", "action": "store_true" },
}
parser = Parser(options)
location = parser.location  # Experiments are stored here
experiments = parser.parse()
parser.document_settings()  # Save a config file to reproduce the experiment
# Run each experiment
for args in experiments:
    run_experiment(location, args)

# Alternatively, if there is only ever a single job
parser = Parser(options, multiple_jobs=False)
location = parser.location
args = parser.parse()
parser.document_settings()
run_experiment(location, args)

# Check if an argument has been given explictly, either from cli or config file, or if default value is used
parser.is_explicit("learning-rate")
```

This could then by run by
`python main.py data/my-big-experiment --learning-rate 1e-5`
or by
`python main.py data/my-big-experiment --config cfg.ini`
or using a combination where CLI args takes precedence:
`python main.py data/my-big-experiment --config cfg.ini --learning-rate 1e-5`
where `cfg.ini` could contain

```
[DEFAULT]
gamma = 0.95

[RUN1]
learning-rate = 1e-4
initialize-zeros

[RUN2]
learning-rate = 1e-5
gamma = 0.9
```

# pelutils.ds

This submodule contains various utility functions for data science and machine learning. To make sure the necessary requirements are installed, install using
```
pip install pelutils[ds]
```
Note that in some terminals, you will instead have to write
```
pip install pelutils\[ds\]
```

## PyTorch

All PyTorch functions work independently of whether CUDA is available or not.

```py
# Inference only: No gradients should be tracked in the following function
# Same as putting entire function body inside with torch.no_grad()
@no_grad
def infer():
    <code that includes feedforwarding>

# Feed forward in batches to prevent using too much memory
# Every time a memory allocation error is encountered, the number of batches is doubled
# Same as using y = net(x), but without risk of running out of memory
# Gradients are not tracked
bff = BatchFeedForward(net)
y = bff(x)
```

## Statistics

Includes various commonly used statistical functions.

```py
# Get one sided z value for exponential(lambda=2) distribution with a significance level of 1 %
zval = z(alpha=0.01, two_sided=False, distribution=scipy.stats.expon(loc=1/2))

# Get correlation, confidence interval, and p value for two vectors
a, b = np.random.randn(100), np.random.randn(100)
r, lower_r, upper_r, p = corr_ci(a, b, alpha=0.01)
```

## Matplotlib

Contains predefined rc params, colours, and figure sizes.

```py
# Set wide figure size
plt.figure(figsize=figsize_wide)

# Use larger font for larger figures - works well with predefined figure sizes
update_rc_params(rc_params)

# 15 different, unique colours
c = iter(colours)
for i in range(15):
    plt.plot(x[i], y[i], color=next(c))
```



HISTORY.md
# History

## 0.6.9 - Nice

- Made `load_jsonl` load the file lazily

## 0.7.0 - Breaking changes

- Removed `log.tqdm`
- Added binning functions to `pelutils.ds.plot`
- Added `pelutils.ds.distributions` with `scipy` distributions that use same notation as Jim Pitman's "Probability"
- Added `examples` directory

  Currently only has plotting examples
- Removed `subfolder` attribute on `DataStorage` as this was cause for much confusion and many issues
- Made naming of different submodules consistent
- Added `linecounter` entry point that allows plotting development of code in repositories over time
- Made `get_repo` return the absolute repository path
- Made logger use absolute instead of relative paths preventing issues with changing working directory
- Renamed `vline` to `hline` in `Table` as the line was horizontal rather than vertical
- Added `jsonl.dumps` and `jsonl.loads` and made all functions in `jsonl` module use same naming scheme as `json`
- Added support for Pytorch tensors to `c_ptr`
- Added `log.no_log` for disabling logging
- Renamed `Levels` to `LogLevels` as that makes it clearer in usage context
- `unique` can now take non-contiguous arrays and also has an `axis` argument
- Added a command line entry to examples, `pelexamples`
- Added a `binary_search` for searching in ordered iterables in `log(n)` time
- Added moving average function and variations thereof for plotting noisy data with uneven spacing
- Made `TickTock.tick` void and `TickTock.tock` raise `TickTockException` if `.tick` not called first
- Changed to `rapidjson` instead for built-in `json` module for all json operations for strictness and that sweet performance
- Added `reverse_line_iterator` for iterating through a file backwards
- Renamed `throws` to `raises`
- Reanedm `thousand_seps` to `thousands_seperators`

### Bug fixes

- Fixed a bug where a backslash would sometimes be printed from the logger before square brackets
- Fixed `raises` throwing errors if error was not caught
- Made `thousands_seperators` work with negative numbers

## 0.6.7

- Logger can now be used without writing to file

## 0.6.6

- Fixed parser naming when using config files and not `multiple_jobs`
- Fixed parser naming when using cli only and `multiple_jobs`

## 0.6.5 - Breaking changes

- `Parser.parse` now returns only a single experiment dict if `multiple_jobs` is False
- Improved logger error messages
- Added `Parser.is_explicit` to check if an argument was given explicitly, either from CLI or a config file
- Fixed bug in parser, where if a type was not given, values from config files would not be used
- Made fields that should not be used externally private in parser
- Made `pelutils.ds.unique` slightly faster

## 0.6.4 - Breaking changes

- Commit is now logged as `DEBUG`
- Removed `BatchFeedForward.update_net`
- `BatchFeedForward` no longer requires batch size and increase factor as an argument
- Removed `reset_cuda` function, as was a too small and obscure function and broke distributed training
- Added `ignore_missing` field to `DataStorage` for ignoring missing fields in stored data

## 0.6.3 - Breaking changes

- Fixed bug where TickTock profiles would sometimes not be printed in the correct order
- Removed `TickTock.reset`
- Added `__len__` and `__iter__` methods to `TickTock`
- Added option to print standard deviation for profiles
- Renamed `TimeUnit` to `TimeUnits` to follow `enum` naming scheme
- Time unit lengths are now given in units/s rather than s/unit

## 0.6.2

- `TickTock.__str__` now raises a `ValueError` if profiling is still ongoing to prevent incorrect calculations
- Printing a `TickTock` instance now indents percentage of time spent to indicate task subsets

## 0.6.1

- Added `subfolder` argument to `Parser.document_settings`

## 0.6.0 - Breaking changes

- A global instance of `TickTock`, `TT`, has been added - similar to `log`
- Added `TickTock.profile_iter` for performing profiling over a for loop
- Fixed wrong error being thrown when keyboard interrupting within `with TT.profile(...)`
- All collected logs are now logged upon an exception being thrown when using `log.log_errors` and `collect_logs`
- Made `log.log_errors` capable of handling chained exeptions
- Made `log.throw` private, as it had little use and could be exploited
- `get_repo` no longer throws an error if a repository has not been found
- Added utility functions for reading and writing `.jsonl` files
- Fixed incorrect `torch` installations breaking importing `pelutils`

## 0.5.9

- Add `split_path` function which splits a path into components
- Fix bug in `MainTest` where test files where not deleted

## 0.5.7

- Logger prints to `stderr` instead of `stdout` at level WARNING or above
- Added `log.tqdm` that disables printing while looping over a `tqdm` object
- Fixed `from __future__ import annotations` breaking `DataStorage`

## 0.5.6

- DataStorage can save all picklable formats + `torch.Tensor` specifically

## 0.5.5

- Test logging now uses `Levels.DEBUG` by default
- Added `TickTock.fuse_multiple` for combining several `TickTock` instances
- Fixed bugs when using multiple `TickTock` instances
- Allow multiple hits in single profile
- Now possible to profile using `with` statement
- Added method to logger to parse boolean user input
- Added method to `Table` for adding vertical lines manually

## 0.5.4 - Breaking changes

- Change log error colour
- Replace default log level with print level that defaults to `Levels.INFO`

  `__call__` now always defaults to `Levels.INFO`
- Print microseconds as `us` instead of `mus`

## 0.5.3

- Fixed missing regex requirement

## 0.5.2

- Allowed disabling printing by default in logger

## 0.5.1

- Fixed accidental rich formatting in logger
- Fixed logger crashing when not configured

## 0.5.0 - Breaking changes

- Added np.unique-style unique function to `ds` that runs in linear time but does not sort
- Replaced verbose/non-verbose logging with logging levels similar to built-in `logging` module

pelutils/parse.py
from __future__ import annotations
import os, sys
from argparse import ArgumentParser, RawTextHelpFormatter
from configparser import ConfigParser
from typing import Any

from pprint import pformat


class Parser:
    """
    Maintains multiple parsers to allow a workflow of mixing use of config files and CLI arguments.

    Can read from single .ini file and CLI arguments. CLI arguments overwrite settings in all experiments defined if ini file.
    In .ini file, defaults for all runs can be set in [DEFAULT] section and multiple runs can be added with their own section.

    The parser returns a list of dicts of the parsed settings for the experiments.

    Quick example:
    A file `main.py` could contain:
    ```
    options = {
        "location": { "default": "local_train", "help": "save_location", "type": str },
        "learning-rate": { "default": 1.5e-3, "help": "Controls size of parameter update", "type": float },
        "gamma": { "default": 1, "help": "Use of generator network in updating", "type": float },
        "initialize-zeros": { "help": "Whether to initialize all parameters to 0", "action": "store_true" },
    }
    parser = Parser(options)
    experiments = parser.parse()

    ```
    This could then by run by
    `python main.py data/my-big-experiment --learning_rate 1e-5`
    or by
    `python main.py data/my-big-experiment --config cfg.ini`
    where `cfg.ini` could contain
    ```
    [DEFAULT]
    gamma = 0.95
    [RUN1]
    learning-rate = 1e-4
    initialize-zeros
    [RUN2]
    learning-rate = 1e-5
    gamma = 0.9
    ```
    """

    _with_config: bool
    location: str
    explicit_args: list[set[str]]

    def __init__(
        self,
        options: dict,
        name = "Experiment",
        description = "Run experiments with these options",
        show_defaults = True,
        description_last = False,
        multiple_jobs = True,  # If this is True, all jobs are put in subfolders
    ):
        """
        Receives a dict of options.
        This should be a dict of dicts where each dict corresponds to an option with the key as the name.
        In the dict for each option, the key "default" should store the default value and all other keys should correspond to
        kwargs in ArgumentParser.add_argument.
        """
        self.options = options
        self._bool_opts = { argname: settings["action"] == "store_false" for argname, settings
            in options.items() if settings.get("action") in ("store_false", "store_true") }
        self._defaults = dict()  # { argname: default value }
        self.name = name
        self.multiple_jobs = multiple_jobs

        # Main parser for CLI arguments
        self._argparser = ArgumentParser(
            description = description,
            formatter_class = RawTextHelpFormatter,
        )
        self._argparser.add_argument("location", help="Location of output", type=str)
        self._argparser.add_argument("-c", "--config",
            help="Location of configuration file to use (if any). Config file should follow .ini format.", metavar="FILE")
        if description_last:
            self._argparser.epilog = description
            self._argparser.description = None

        abbrvs = {"-h", "-c"}  # -h is reserved for --help and -c for --config
        for argname, settings in self.options.items():
            self._defaults[argname] = settings.pop("default") if argname not in self._bool_opts else self._bool_opts[argname]

            if show_defaults and "help" in settings:
                settings["help"] += f"\n  Default = {self._defaults[argname]}"

            # Add argument and optionally abbrevation if no conflict
            abbrv = f"-{argname[0]}"
            if abbrv in abbrvs:
                self._argparser.add_argument(f"--{argname}", **settings)
            else:
                self._argparser.add_argument(abbrv, f"--{argname}", **settings)
                abbrvs.add(abbrv)

        # Parser for config file
        self._configparser = ConfigParser(allow_no_value=True)

    def _parse_known_args(self) -> dict[str, Any]:
        """ Returns a dict containing the arguments given explicitly from the command line """
        args, __ = self._argparser.parse_known_args()
        args = vars(args)
        known_args = dict()
        for argname, value in args.items():
            if value is None or (argname in self._bool_opts and self._bool_opts[argname] == value):
                continue
            known_args[argname] = value
        return known_args

    def parse(self) -> list[dict[str, Any]] | dict[str, Any]:
        """ Parse arguments and return a dict for each. Only a single dict is returned if multiple_jobs is False """
        # Parse command line arguments
        args = self._parse_known_args()
        self.location = args["location"]
        explicit_cli_args = set(args)

        # Parse config files
        experiments, explicit_config_args = self._read_config(args)
        self._with_config = bool(experiments)

        if not self._with_config:  # If CLI arguments only
            args = { **self._defaults, **args }
            # Allow us to overwrite the hard coded name by adding a `name` CLI arg
            name = args.get("name", self.name)
            if self.multiple_jobs:
                args["location"] = os.path.join(self.location, name)
            experiments.append({"name": name, **args})
            self.explicit_args = [explicit_cli_args]
        else:
            self.explicit_args = [set.union(explicit_cli_args, conf_args) for conf_args in explicit_config_args]

        # Replace - with _ as argparse also does. This allows parsing experiments to a function using **
        for i in range(len(experiments)):
            experiments[i] = { kw.replace("-", "_"): v for kw, v in experiments[i].items() }
            self.explicit_args[i] = { arg.replace("-", "_") for arg in self.explicit_args[i] }

        return experiments if self.multiple_jobs else experiments[0]

    def _set_bools_in_dict(self, d: dict[str, Any]):
        """ Boolean arguments present are set to the negation of their default values
        Those not present are set to default values """
        for argname, default_value in self._bool_opts.items():
            if argname in d and not isinstance(d[argname], bool):
                d[argname] = not default_value
            else:
                d[argname] = default_value

    def _read_config(self, cli_args: dict) -> tuple[list[dict], list[set]]:
        """ Parses a configuration file. Options in cli_args override config files
        Returns a list of experiments and a list of explicitly given config arguments """
        experiments = list()
        explicit_config_args = list()

        if "config" in cli_args:
            if not self._configparser.read([cli_args["config"]]):
                raise FileNotFoundError(f"Could not find config file {cli_args['config']}")

            # User set DEFAULT section should overwrite the defaults
            default_config_items = dict(self._configparser.items("DEFAULT"))
            self._defaults = { **self._defaults,  **default_config_items }

            if len(self._configparser) > 1 and not self.multiple_jobs:
                raise ValueError(
                    "Multiple jobs are given in the config file, however the parser has been configured for a single job"
                )

            # Each other section corresponds to an experiment
            for experiment_name in self._configparser.sections() if self._configparser.sections() else ["DEFAULT"]:
                config_items = dict(self._configparser.items(experiment_name))
                explicit_config_args.append(
                    set.union(set(default_config_items), config_items)
                )
                config_items = {
                    kw: self.options[kw]["type"](v) if "type" in self.options[kw] else v
                    for kw, v in config_items.items()
                }
                options = { **self._defaults, **config_items }
                self._set_bools_in_dict(options)

                experiment_name = experiment_name if experiment_name != "DEFAULT" else self.name

                # Put experiments into single subfolders
                location = os.path.join(self.location, experiment_name)\
                    if self.multiple_jobs else self.location

                args = cli_args.copy()
                del args["config"]
                experiments.append({
                    **options,
                    **args,
                    "name": experiment_name,
                    "location": location,
                })

        return experiments, explicit_config_args

    def document_settings(self, subfolder=""):
        """ Saves all settings used for experiments for reproducability """
        os.makedirs(os.path.join(self.location, subfolder), exist_ok = True)

        with open(os.path.join(self.location, subfolder, self.name + "_config.ini"), "w") as f:
            if self._with_config:
                self._configparser.write(f)
            f.write(f"\n# Run command\n# {' '.join(sys.argv)}\n")
            str_defaults = pformat(self._defaults).replace("\n", "\n# ")
            f.write(f"\n# Default configuration values at runtime\n# {str_defaults}")

    def is_explicit(self, argname: str, job: int=None) -> bool:
        """ Checks whether a given argument was set explicitly in a config file or from cli
        If self.multiple_jobs is False, job should not be given. Otherwise, a job number must be given """
        return argname in self.explicit_args[job if self.multiple_jobs else 0]
